addons/purchase/models/account_invoice.py

```diff
@@ -148,119 +149,312 @@ def write(self, vals):
+                # We return all purchase order lines of the purchase order whose total amount matched our vendor bill.
+                return ('total_match',
+                        matching_purchase_orders.order_line,
+                        None)
+
+        # We couldn't find anything, so we return no lines.
+        return ('no_match', matching_purchase_orders.order_line, None)
+
+    def _find_and_set_purchase_orders(self, po_references, partner_id, amount_total, from_ocr=False, timeout=10):
+        """Finds related purchase orders that (partially) match the vendor bill and links the matching lines on this
+        vendor bill.
+
+        :param po_references: a list of potential purchase order references/names
+        :param partner_id: the vendor id matched on the vendor bill
+        :param amount_total: the total amount of the vendor bill
+        :param from_ocr: indicates whether this vendor bill was created from an OCR scan (less reliable)
+        :param timeout: the max time the line matching algorithm can take before timing out
+        """
         self.ensure_one()
 
-        method, matched_po_lines = self._match_purchase_orders(po_references, partner_id, amount_total, timeout)
+        method, matched_po_lines, matched_inv_lines = self._match_purchase_orders(po_references, partner_id, amount_total, from_ocr, timeout)
 
-        if method == 'total_match': # erase all lines and autocomplete
+        if method in ('total_match', 'po_match'):
+            # The purchase order reference(s) and total amounts match perfectly or there is only one purchase order
+            # reference that matches with an OCR invoice. We replace the invoice lines with the purchase order lines.
             self._set_purchase_orders(matched_po_lines.order_id, force_write=True)
 
-        elif method == 'subset_total_match': # don't erase and add autocomplete
+        elif method == 'subset_total_match':
+            # A subset of the referenced purchase order lines matches the total amount of this invoice.
+            # We keep the invoice lines, but add all the lines from the partially matched purchase orders:
+            #   * "naively" matched purchase order lines keep their quantity
+            #   * unmatched purchase order lines are added with their quantity set to 0
             self._set_purchase_orders(matched_po_lines.order_id, force_write=False)
 
-            with self._get_edi_creation() as move_form: # logic for unmatched lines
+            with self._get_edi_creation() as move_form:
                 unmatched_lines = move_form.invoice_line_ids.filtered(
                     lambda l: l.purchase_line_id and l.purchase_line_id not in matched_po_lines)
                 for line in unmatched_lines:
-                    if prefer_purchase_line:
-                        line.quantity = 0
-                    else:
-                        line.unlink()
+                    line.quantity = 0
```

Comment:
update all quantities at once using orm `Command.update` like
```suggestion
invoice.invoice_line_ids = [
    Command.update(line.id, {'quantity': 0})
    for line in unmatched_lines
]
```

Reply:
Good you noticed. I didn't think of this possibility before.